---
title: "MultiIP Reports"
author: "Yu Sun"
date: "June 26, 2017"
# output: html_document
output:
  flexdashboard::flex_dashboard:
    navbar:
      - { title: "About", href: "https://icannotendure.space/", align: right }
      - { icon: "ion-social-github", href: "https://github.com/bioinformatist/multiIP", align: right }
    theme: flatly
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, fig.width = 9, fig.height = 6, cache = FALSE, message = FALSE, warning=FALSE)
rm(list = ls())
gc()
# The latest plotly should be used, for a lot of bugs fixed now
# if (!require("devtools")) install.packages("devtools")
# devtools::install_github("ropensci/plotly")
require(plotly)
require(data.table)
require(cowplot)
require(stringr)
require(knitr)
.toupperFirstLetter = function(x) paste0(toupper(substr(x, 1, 1)), tolower(substring(x, 2)))
```

Quality Control
===================================== 

Column
-------------------------------------

### *FastQC* results

```{r results='asis'}
fastqc.dirs = list.files(pattern = '_fastqc$')
fastqc.table = data.table(Sample = substr(fastqc.dirs, 7, nchar(fastqc.dirs)-7), HTML = paste0('[', substr(fastqc.dirs, 7, nchar(fastqc.dirs)-7), '](', paste0(fastqc.dirs, "/fastqc_report.html"), ')'), Summary = paste0('[', substr(fastqc.dirs, 7, nchar(fastqc.dirs)-7), '](', paste0(fastqc.dirs, "/summary.txt"), ')'), Details = paste0('[', substr(fastqc.dirs, 7, nchar(fastqc.dirs)-7), '](', paste0(fastqc.dirs, "/fastqc_data.txt"), ')'))
kable(fastqc.table, caption = "FastQC results")
rm(fastqc.dirs, fastqc.table); invisible(gc())
```

STAR {data-navmenu="Aligning"}
===================================== 

```{r}
# Call perl one-liner to concatenate all log files in the list
log.summary = fread(
paste0(
"grep '' `",
"cat log.list",
"` | perl -F':\\s{2,}|\\s\\|\\t' -lanE'/:$/ ? next : say join qq{\t}, @F'"
),
header = FALSE
)
# Extract lines and get sample names
log.summary <-
log.summary[V2 %like% c('of i|Uni|mu|many l|sho|oth')] [, V1 := tstrsplit(V1, '\\.')[[1]]]
# Convert percentage (characters) into decimals
log.percentage <-
log.summary[V2 %like% '%'][, ':=' (V2 = .toupperFirstLetter(gsub('% of reads | reads %', '', V2)), V3 = as.numeric(sub("%", "", V3, fixed = TRUE)) / 100)]
setnames(log.percentage, c('Sample', 'Type', 'Percentage'))
```

Column 
-------------------------------------

### *STAR* log table

```{r}
DT::datatable(log.percentage)
```

Column 
-------------------------------------

### *STAR* log plot

```{r}
p <-
ggplot() + geom_bar(data = log.percentage,
aes(x = Sample, y = Percentage, fill = Type),
stat = 'identity') + scale_y_continuous(labels = scales::percent) + coord_flip()
# p
ggplotly(p)
rm(log.summary, log.percentage, p)
invisible(gc())

##############################################################################################################################
# system.time(
#   log.list[V2 %like% c('Uni')]
# )
#  user  system elapsed
# 0.000   0.000   0.001

# system.time(
#   log.list[V2 %in% c('Uniquely mapped reads number', 'Uniquely mapped reads %')]
# )
#  user  system elapsed
# 0.003   0.000   0.003
##############################################################################################################################
```

HISAT2 {data-navmenu="Aligning"}
=====================================

Developing...

CDF {data-navmenu="Annotation"}
=====================================

```{r}
cdf <- fread('CDF.txt')
# Order first, for using geom_line later (geom_step behaves badly when calling plotly, see https://github.com/ropensci/plotly/issues/1030)
cdf <- cdf[ order(cdf[,1], cdf[,2]) ]
# Subset data randomly
cdf <- cdf[sample(nrow(cdf), nrow(cdf) * 0.25), ]
setnames(cdf, c('Type', 'CPAT'))
```

Column
-------------------------------------

### *CPAT* table

```{r}
DT::datatable(cdf)
```

Column
-------------------------------------

### *CDF* against *CPAT*

```{r}
p <- ggplot() + geom_line(data = cdf, aes(x = CPAT, colour = Type), size = 2, stat = 'ecdf') + scale_x_continuous(expand  = c(.01, 0)) + scale_y_continuous(expand = c(0, 0)) + labs(x = 'Coding Probablity(CPAT)', title = "Coding Potential", y = "CDF") + scale_color_manual(values = c("#FFA500", "#3A5FCD", "#EE2C2C")) + geom_hline(yintercept = 1, colour = "grey",linetype = "dashed",size = 1)
ggplotly(p)
rm(cdf, p); invisible(gc())
```

Transcript length {data-navmenu="Annotation"}
=====================================

```{r}
gtf <- fread(paste("sed -re 's/\\s+|;\\s+|;/\t/g'", "lncRNA.final.v2.gtf", "| cut -f2,4,5,12"))
# Use sum of exon length as transcript length
gtf <- unique(gtf[, V2 := sum(V3-V2), by = .(V1, V4)][, V3 := NULL])
setcolorder(gtf, c('V4', 'V1', 'V2'))
setnames(gtf, c('Transcript', "Type", "Length"))
```

Column
-------------------------------------

### Transcript length table

```{r}
DT::datatable(gtf)
```

Column
-------------------------------------

### Transcript length density plot

```{r}
p <- ggplot() + geom_density(data = gtf, aes(x = Length, colour = Type)) + xlab('Transcript length') + ylab('Density') + scale_x_continuous(expand = c(0.01, 0)) + scale_y_continuous(expand = c(0.01, 0))
ggplotly(p)
rm(gtf, p); invisible(gc())
```


